### ISSUE TITLE
Add a data source for storage accounts

### ISSUE THREAD

**Issue by eefi** [terraform-providers/terraform-provider-azurerm#785]
We have a use case that could really make use of a storage account data source. Is there a philosophical reason why that doesn't exist right now? (Or why the current set of data sources are the data sources that do exist?)

We have our infrastructure deployed through a collection of modules, and one of our patterns has been to define a shared resource in one module, and then use a data source in another module to do things to it. For example, a core module might define a virtual network resource, and then we'd make a data source for that virtual network in a different module that might want to add a subnet to it.

I understand the data source isn't strictly necessary, since you can generally just refer to things by name, which you need to get the data source in the first place anyway. But it's a little peace of mind (and easier debugging?) when the data source object fails first during a Terraform run if the referenced resource isn't there.

**Comment by tombuildsstuff** [comment]
hey @eefi

Thanks for opening this issue :)

> We have a use case that could really make use of a storage account data source. Is there a philosophical reason why that doesn't exist right now? (Or why the current set of data sources are the data sources that do exist?)

Totally agree a Data Source for Storage Accounts could be really useful - as such I've opened PR #794 which adds support for this.

Historically Data Sources have been added as the demand for them arises, rather than adding them ahead of time. There's a longer term task to go through and backfill some of the existing Resources as Data Sources here (e.g. Storage Accounts) - but for new Resources we're adding a Data Source shortly after the Resource is supported if it makes sense (for instance, Network Security Groups probably will get a Data Source, but I don't believe we need one for Network Security Rules [since you're not likely to reference it elsewhere]).

Thanks!
